diesel/src/types/mod.rs

```diff
@@ -274,10 +275,10 @@ pub type VarChar = Text;
 pub use pg::types::sql_types::*;
 
 pub trait HasSqlType<ST>: TypeMetadata {
-    fn metadata() -> Self::TypeMetadata;
+    fn metadata(lookup: &Self::MetadataLookup) -> Self::TypeMetadata;
```

Comment:
I implemented this and compared the two versions. I decided that I did not like the enum form. Reasoning laid out in the commit message. https://github.com/diesel-rs/diesel/commit/2a84869c38b0d69bddf86021029e5c003c7e74f0